### DISCUSSION TITLE
Partial pattern matches when using operators and quantifiers

### DISCUSSION THREAD

**Discussion by databill86** [explosion/spaCy#6539] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour
Here is my testing pattern:
`patterns = [
  [
    { "LOWER": { "IN": ["la", "le", "les"] } },
    { "LOWER": "rendre" },
    { "OP": "?" },
    { "LOWER": { "NOT_IN": ["court", "pratique"] } }
  ]
]`

I'm looking for texts like: '_le rendre plus X_', where X can be anything but '_court_' or '_pratique_'.
`
from spacy.matcher import Matcher
matcher = Matcher(nlp.vocab)
for idx, pattern in enumerate(patterns):
    matcher.add(f"pat-{idx}", None, pattern)

text = "le rendre plus pratique"
doc = nlp(text)

matches = matcher(doc)
print(matches)
`

We expect the matcher to do not return any match from the text. However, this is the result:

`[(2241456119784813780, 0, 3)] => [{'start': 0, 'end': 14}]`
which means that it partially matched the sequence _le rendre plus_ and ignored the last condition. 

Is this the expected behavior?

I tried another pattern using the '!' quantifier:

`patterns = [    [
     {'LOWER': {'IN': ['la', 'le', 'les']}}, 
     {'LOWER': 'rendre'},
    { 'OP': '!', 'LOWER': {'IN': ['très']}},
    {'LOWER': {'NOT_IN': ['court', 'pratique']}}]]`

The idea is to get sequences like _le rendre X Y_ where X can't be _très_ and Y can't be in `['court', 'pratique']`.
The matcher fails to get a match for this text: _le rendre plus fin_. It does not match _le rendre très pratique_ neither, which is expected.

Is this a bug when using quantifiers and operators along with 'IN' or 'NOT_IN'?

## Your Environment

* Operating System: Unix
* Python Version Used: 3.7
* spaCy Version Used: 2.3.2
* Environment Information: JupyterLab on Docker, miniconda3:4.8.2

**Comment by adrianeboyd** [comment, marked as answer]
Yes, this looks like the expected behavior. What's going on is that `{"OP": "?"}` is not matching any token (it is allowed to match 0 tokens) and then `plus` is matching `{"LOWER": {"NOT_IN": ["court", "pratique"]}}`.

**Comment by databill86** [comment]
Do you have any suggestions in order to get only the matches I'm looking for ?

What about the second pattern, when using '!'? 

Thank you!
